File: frontend/src/shared/lib/files/limits.js
/**
 * How big a file may be, said before somebody spends a minute finding out.
 *
 * There was no size messaging anywhere near an upload control. The only
 * sentence in the SPA about a size was `FileSurface`'s "too big to show here",
 * which is about *reading* a file. A person picked a 40 MB video, watched it
 * travel, and was told at the end.
 *
 * The number comes from the server, because it is not the same on every site:
 * below `directUpload.THRESHOLD` a file is POSTed to Frappe and the
 * framework's `max_file_size` bounds it; above it the bytes go straight to R2
 * in parts and nothing in between reads the body. So on a site with a bucket
 * there is no per-file ceiling worth printing and this says nothing — the
 * thing that refuses a large file there is the quota, which has its own words
 * and its own number. `onestorage/limits.py` decides which case a site is in.
 *
 * `docs/UNIFICATION.md` §D3.
 */
import { limits } from '@/shared/lib/runtime/boot'
import { sizeText } from '@/shared/lib/files/size'
import { __ } from '@/shared/lib/runtime/translate'

/** Bytes, or 0 where there is no fixed per-file ceiling here. */
export const CEILING = Number(limits?.file) || 0

/**
 * The sentence that goes under an attach control, or '' when there is no
 * ceiling to state. One clause, because it sits beside the control rather
 * than instead of it.
 */
export function ceilingNote() {
  return CEILING ? __('Up to {0} each.', [sizeText(CEILING)]) : ''
}

/** Whether this one would be refused. */
export const oversize = (file) => !!CEILING && Number(file?.size || 0) > CEILING

/**
 * What of this may be sent, and why the rest may not.
 *
 * Checked here rather than left to the server, because the server's refusal
 * arrives after the bytes did. Returns the sentence rather than raising it: a
 * dialog wants it in its `ErrorMessage` and a drop target wants it in a toast,
 * and neither wants an exception.
 */
export function withinCeiling(files) {
  const list = Array.from(files || [])
  const good = list.filter((one) => !oversize(one))
  if (good.length === list.length) return { good, refused: [], why: '' }

  const refused = list.filter(oversize)
  return {
    good,
    refused,
    why:
      refused.length === 1
        ? __('{0} is {1}, and the largest that can go here is {2}.', [
          refused[0].name, sizeText(refused[0].size), sizeText(CEILING),
        ])
        : __('{0} files are larger than {1} and were not sent.', [
          refused.length, sizeText(CEILING),
        ]),
  }
}

File: oneapp/onestorage/limits.py
"""How big a file may be, here, and who is allowed to say so.

There were two ceilings and neither was ever said out loud. WebDAV has one it
enforces and explains — `dav.py` returns a whole sentence with a 413 — and the
browser has one it does not mention at all: a person picks a 40 MB file, waits
for it to travel, and is told at the end that it was too big. The only size
messaging anywhere in the SPA was `FileSurface`'s "too big to *show* here",
which is about reading.

The number is not the same on every site, which is why it has to be asked for
rather than written down. Below `direct.THRESHOLD` a file is POSTed to Frappe
and the framework's own `max_file_size` bounds it. Above it the browser sends
the bytes straight to R2 in parts, and nothing between here and Cloudflare
reads the body — so on a site with a bucket there is no per-file ceiling worth
printing, and the thing that actually refuses is the quota, which says so in
its own words.

So: a number when there is one, and zero when there is not.

`docs/UNIFICATION.md` §D3.
"""

import frappe
from frappe.utils import cint

from oneapp.onestorage import r2
from oneapp.onespace import site

#: Frappe's own default, from `init_request`. Named here because the config key
#: is usually absent and 25 MB is then the real answer rather than "unset".
FRAMEWORK_DEFAULT = 25 * 1024 * 1024


def posted_ceiling() -> int:
	"""The largest body Frappe itself will accept on this site.

	Frappe's own expression, from `init_request`, and deliberately not
	`frappe.core.api.file.get_max_file_size` — that one consults System
	Settings first, but only for `/api/method/upload_file`, so quoting it on
	any other path names a number that does not apply.
	"""
	return cint(frappe.local.conf.get("max_file_size")) or FRAMEWORK_DEFAULT


def browser_ceiling() -> int:
	"""The largest single file a browser may put in, or 0 for no fixed one.

	Zero wherever the direct path is available, because then the bytes never
	pass through the framework and the only thing that refuses a large file is
	the quota — which has its own message and its own number.
	"""
	if site.is_control() or not r2.is_configured():
		return posted_ceiling()
	return 0


def boot() -> dict:
	"""What the SPA needs before it draws an attach control.

	Before first paint rather than on a resource, because the sentence belongs
	*under the control* — a limit a person reads after choosing the file is a
	limit they have already spent a minute on.
	"""
	return {"file": browser_ceiling()}
